### DISCUSSION TITLE
Option to display Scripts and Docs as tabs, above the Text Editor

### DISCUSSION THREAD

**Discussion by Mickeon** [godotengine/godot-proposals#4248] [Editor]
A while ago I wrote a proposal on adding an [**option** to the Script Editor to have it display Scripts and Docs as tabs](https://github.com/godotengine/godot-proposals/issues/4081), making it more similar to other IDE's.

Here's the same mockup used in the proposal, for easy access:
![image](https://user-images.githubusercontent.com/66727710/155152644-c7261d61-6f03-458a-88f8-98da8d50a00a.png)

However, its reception was mixed and unexpectedly lackluster. 
It could be that I was too long winded, or I explained my reasoning exceedingly poorly, or that the **optional** feature is **so not** worth maintaining that users straight up dislike it.

By opening this discussion, I would like to squeeze a few extra opinions on why this would be neat or horrible.

**Comment by lentsius-bark** [comment]
Personally I prefer tabs to a list. Sure the user can fit less many of them but I feel that one does not need a that many when working on tasks anyway.
Also I noticed your mockup has two rows, one for docs and the other for scripts, I really like that too.

I feel that the main reason I like tabs is that it forces me to have a lean list scripts I'm working on, which I focus on, rather than a super long list on the side where I far too often just forget about all of the other ones and just use the quick open menu to get to anyway.
Ah and the additional space as well, that's cool

**Comment by Calinou** [comment]
One downside of tabs is that they take vertical space instead of horizontal space. In UIs, vertical space is much more precious than horizontal space, especially on 16:9 displays which are often criticized for lacking vertical space.

This is why I'd prefer keeping the list, and even find other ways to *save* vertical space rather than take it away.

**Reply by AndHalfAFish** [reply]
I feel exactly the opposite way - on my laptop the list takes up valuable horizontal space since there are basically 3 columns : scene/file system, script/method list, and inspector. This isn't a problem w an external monitor, but on my laptop it leaves little space for the script editor ...

**Reply by Zireael07** [reply]
I have the exact same problem if I make Godot window smaller to e.g. see some code I'm using as reference

**Reply by Mickeon** [reply]
> In UIs, vertical space is much more precious than horizontal space, especially on 16:9 displays which are often criticized for lacking vertical space.

It is generally true, yeah, but Godot usually takes up 4 columns (SceneTree & FileSystem, Script List, Text Editor, Inspector). The Script List can be toggled, and a Focus Mode IS present, but due of the columns' importance it'd be nice to fit their functions nicely in a laptop screen, or on a 1080p screen with enough space to just fit another window. I also find that having this information in full display helps me focus as a constant reminder of what I'm working on (probably personal preference).

Overall, I've found horizontal space to be quite more valuable in the Editor than at first assumed, especially as long, clear, self-explanatory variable and method names are encouraged.

> I have the exact same problem if I make Godot window smaller to e.g. see some code I'm using as reference

I'd like to see the code, actually, as a reference for other user experiences.

**Reply by Zireael07** [reply]
@Mickeon: No code here, just a screenshot of me porting some c# code (I don't use Mono Godot, I just use GDScript, in the future I might use Rust when Godot 4 and GDExtension gets stable enough)
![image](https://user-images.githubusercontent.com/4525388/160613083-1a958e15-4000-4e84-a6a3-c60f991edbbf.png)

This is a 24 inch monitor, not some 14" inch laptop btw. Every bit of horizontal space counts, see how narrow my script editor is, I wish I could fold/pop-out inspector and scene tree like some mobile apps do, but moving the script list to horizontal would help. (Maybe this could happen automatically below some horizontal size sort of like web pages are now responding to different window/device sizes?)

**Comment by theraot** [comment]
I think both tabs and and a list of open files makes sense. However, I would put the list of open files in the FileSystem panel, similar to the Favorites category.

However, people use the methods list (in fact, here is the idea to add all members there: https://github.com/godotengine/godot-proposals/discussions/4291). I think that could be a drop down at the top, like Visual Studio has been doing for decades now.

I do also care about vertical space. I would rather documentation and file tabs were on the same level are scene tabs. Although I admit I haven't fully fleshed out my idea yet.

**Comment by Grind-Knight** [comment]
I know this is a bit of an older post but I'm all for this as I recently switched back from VSC to the Godot editor (to reduce the amount of windows open when editing several projects at once) and the lack of tabs really limits my visibility and makes it difficult to navigate between files. Even if there was less visible tabs at one time you could, like other IDE's, just mouseover and scroll to cycle through. Then we'd still be saving plenty of space.

**Comment by testeria** [comment]
Please make this an option, in my job I always worked with VS, and going against a 20-year-old habit is awful. I literary ALWAYS switch scenes instead of files when I'm working in Godot editor - and I mostly work with scripts, not scenes.

**Comment by Gnumaru** [comment]
There's an asset in the Asset Library for showing open scripts and docs as tabs.

https://godotengine.org/asset-library/asset/2080

https://github.com/MakovWait/godot-script-tabs

Personally I preffer tabs on top instead of a list on the side. The problem with script names taking too much horizontal space in tabs and the dificulty of them being viewed without a horizontal scroll bar can be solved in the same way that sublime text and other text editors does: with a dropdown at the right of the tab list that shows a vertical list of all existing tabs.

![image](https://github.com/godotengine/godot-proposals/assets/5783414/f71c475b-00c9-4739-aa5f-47e2d47faba2)

visual studio code does the same thing, but the button actually calls an editor action and the list of files is shown in the center of the screen

![image](https://github.com/godotengine/godot-proposals/assets/5783414/58b57666-037b-498f-80b7-af257e290a64)
